core/modules/vxlan_encap.cc

```diff
@@ -107,9 +107,23 @@ void VXLANEncap::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
     vh->vx_flags = be32_t(0x08000000);
     vh->vx_vni = vni << 8;
 
-    udp->src_port = be16_t(
-        rte_hash_crc(inner_eth, sizeof(Ethernet::Address) * 2, UINT32_MAX) |
-        0xf000);
+    uint32_t h = 0;
+    if (inner_eth->ether_type.value() != Ethernet::Type::kIpv4) {
+      h = rte_hash_crc(inner_eth, sizeof(Ethernet::Address) * 2, UINT32_MAX);
+    } else {
+      Ipv4 *inner_ip = reinterpret_cast<Ipv4 *>(inner_eth + 1);
+      size_t ip_len = inner_ip->header_length;
+      h = inner_ip->protocol;
+      if (inner_ip->protocol == Ipv4::Proto::kTcp ||
+          inner_ip->protocol == Ipv4::Proto::kUdp) {
+        Udp *inner_l4 = reinterpret_cast<Udp *>(
+            reinterpret_cast<uint8_t *>(inner_ip) + ip_len);
+        h |= rte_hash_crc_2byte(inner_l4->src_port.value(),
```

Comment:
Did you intend `h ^= ...`? `|=` will give you more 1's than 0's...

Or alternatively, you can do `h = rte_hash_crc_4byte(&inner_l4->src_port, h);`. It should be faster as it has only one memory access, without `bswap` caused by `value()`.